### PR TITLE
Add TradeExecutionModel - fixes #598

### LINKED ISSUE #598: Results on Backtesting
Hi,

First of all I would like to thank you for all of your effort you spend in this lib - great work.
I have tested the backtesting  - when you run a strategy and analyse tradingRecord i was wondering if you take the right bar for the entry:

 - in the trading records the index of entry is the index where the indicator rule is true (index x)
 - you save the close price(AssetPrice) of the bar x in the trade order

But in general that is not true for live trading:
 - you get bar X in the stream and analyze your indicator and decide to enter the trade
 - then you decide to enter the market and you will get the Open price of bar x+1

Do I get something wrong ? If this is the case then the backtesting results are not correct?

### PR DESCRIPTION
**Pull request by phong-phuong**
Fixes https://github.com/ta4j/ta4j/issues/598 where users expect backtests to open and close trades on the open of the next bar after a confirmed signal.

Changes proposed in this pull request:

Adds a new interface: TradeExecutionModel, which allows trade execution behaviour to be modified
Two implementations of TradeExecutionModels are provided, TradeOnNextOpenModel, and TradeOnCurrentCloseModel
Update BarSeriesManager to use TradeOnNextOpenModel by default if no TradeExecutionModel is specified
Updates existing tests to use TradeOnCurrentCloseModel so logic is not affected
Basic test added to ensure TradeOnNextOpenModel opens on the next bar and at the open price
Moved BarSeriesManagerTest from root package to backtest package

Breaking:

Moved BarSeriesManager and BacktestExecutor from root package to backtest package

### CODE REVIEW

File: ta4j-core/src/main/java/org/ta4j/core/backtest/TradeExecutionModel.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+/**
+ * An execution model for {@link BarSeriesManager} objects.
+ *
+ * Used for backtesting. Instructs {@link BarSeriesManager} on how to execute
+ * trades.
+ * 
+ */
+public interface TradeExecutionModel {
+
+    void apply(int index, TradingRecord tradingRecord, BarSeries barSeries, Num amount);
```
Comment (nimo23) [inline ta4j-core/src/main/java/org/ta4j/core/backtest/TradeExecutionModel.java]: Please add the following javadoc to this interface method:

```
/**
  * Places a trade in the given {@code tradingRecord}.
  * 
  * @param index         the trade index from {@code barSeries} 
  * @param tradingRecord the trading record to place the trade
  * @param barSeries     the bar series 
  * @param amount        the trade amount
  */
void apply(int index, TradingRecord tradingRecord, BarSeries barSeries, Num amount);
```

**Review by team172011 — COMMENTED** [review]
Would it make sense to change the names of the `TradeExecutionModels` to

- `TradeOnCurrentCloseModel`
- `TradeOnNextOpenModel`

?

**Comment by phong-phuong** [comment]
Yes I like those names, and it is short enough to use.

File: CHANGELOG.md [hunk 2]
```diff
@@ -45,6 +51,7 @@ Changelog for `ta4j`, roughly following [keepachangelog.com](http://keepachangel
 - added "lessIsBetter"-property for **StandardErrorCriterion**
 - added "lessIsBetter"-property for **VarianceCriterion**
 - added "lessIsBetter"-property for **NumberOfPositionsCriterion**
+- added "TradeExecutionPolicy" to modify trade execution during backtesting
```
Comment (team172011) [inline CHANGELOG.md]: what is/means this "TradeExecutionPolicy"?

**Review by team172011 — APPROVED** [review]

**Comment by team172011** [comment]
@phong-phuong I would merge this if you resolve conflicting files

**Comment by phong-phuong** [comment]
Yeah I'm fixing it...

**Comment by phong-phuong** [comment]
Good now

File: ta4j-core/src/main/java/org/ta4j/core/backtest/BacktestExecutor.java [hunk 3]
```diff
@@ -50,16 +54,9 @@ public BacktestExecutor(BarSeries series) {
...
     }
 
-    /**
-     * Constructor.
-     *
-     * @param series               the bar series
-     * @param transactionCostModel the cost model for transactions of the asset
-     * @param holdingCostModel     the cost model for holding the asset (e.g.
-     *                             borrowing)
-     */
```
Comment (nimo23) [inline ta4j-core/src/main/java/org/ta4j/core/backtest/BacktestExecutor.java:60]: Please do not remove the javadoc.

File: ta4j-core/src/main/java/org/ta4j/core/backtest/TradeExecutionModel.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import org.ta4j.core.BarSeries;
+import org.ta4j.core.TradingRecord;
+import org.ta4j.core.num.Num;
+
+/**
+ * An execution model for {@link BarSeriesManager} objects.
+ *
+ * Used for backtesting. Instructs {@link BarSeriesManager} on how to execute
+ * trades.
+ *
```
Comment (nimo23) [inline ta4j-core/src/main/java/org/ta4j/core/backtest/TradeExecutionModel.java]: Remove the _extra line_ after "trades."

Comment (phong-phuong) [reply ta4j-core/src/main/java/org/ta4j/core/backtest/TradeExecutionModel.java]: The formatter adds the extra line because it's too long

Comment (nimo23) [reply ta4j-core/src/main/java/org/ta4j/core/backtest/TradeExecutionModel.java]: No, I mean this:

```
/**
 * An execution model for {@link BarSeriesManager} objects.
 *
 * Used for backtesting. Instructs {@link BarSeriesManager} on how to execute trades.
 * Used for backtesting. Instructs {@link BarSeriesManager} on how to execute
 * trades.
 *  > DELETE THIS EXTRA BLANK LINE
 */
```

**Review by team172011 — APPROVED** [review]

**Pull request merged**